src/test/java/bio/terra/service/auth/iam/IamServiceTest.java

```diff
@@ -244,4 +252,58 @@ void testVerifyResourceTypeAdminAuthorizedFalse() throws InterruptedException {
             iamService.verifyResourceTypeAdminAuthorized(
                 TEST_USER, IamResourceType.DATASNAPSHOT, IamAction.ADMIN_READ_SUMMARY_INFORMATION));
   }
+
+  @Test
+  void testConstructFirecloudGroupName() {
+    String id = "123456";
+    assertEquals(iamService.constructFirecloudGroupName(id), String.format("%s-users", id));
+  }
+
+  @Test
+  void testConstructUniqueFirecloudGroupName() {
+    String id = "123456";
+    String actual = iamService.constructUniqueFirecloudGroupName(id);
+    // we don't know exactly what our backstop name will be as it will be suffixed by a new UUID,
+    // but we know that it will start with our more readable group name.
+    assertTrue(actual.contains(String.format("%s-users-", id)));
+    assertEquals(49, actual.length());
+  }
+
+  @Test
+  void testCreateFirecloudGroup() {
+    iamService = spy(iamService);
```

Comment:
We should avoid `spy()` if possible. Wouldn't mocking `IamProvider` be sufficient here?

Reply:
I want to mock calls to createGroup on the iamService while testing createFirecloudGroup. So, I'm not sure how just mocking the IamProvider would allow me to do this. Maybe I could make this clearer by renaming createFirecloudGroup -> createGroupWithRetry? That would be more specific and capture the functionality better.